File: 01_arrays_and_hashing/06_product_of_array_except_self/solution.py
from typing import List


class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        result = [1] * n

        right_pass = 1
        for i in range(n):
            result[i] = right_pass
            right_pass *= nums[i]
            print(right_pass, result, nums[i])

        left_pass = 1
        for i in range(n - 1, -1, -1):
            result[i] *= left_pass
            left_pass *= nums[i]
            print(left_pass, result, nums[i])

        return result


if __name__ == "__main__":
    a = [1, 2, 3, 4]
    s = Solution()
    print(s.productExceptSelf(a))

File: 01_arrays_and_hashing/06_product_of_array_except_self/solution.go
package main

func productExceptSelf(nums []int) []int {
	result := make([]int, len(nums))
	result[0] = 1
	for i := 1; i < len(nums); i++ {
		result[i] = result[i-1] * nums[i-1]
	}

	right := 1
	for i := len(nums) - 1; i >= 0; i-- {
		result[i] *= right
		right *= nums[i]
	}

	return result
}

func main() {
	nums := []int{1, 2, 3, 4}
	result := productExceptSelf(nums)
	for _, v := range result {
		println(v)
	}
}
